syntax = "proto3";

import "ca_protomsg.proto";


message GetSDKReq
{
    string   version = 1;
    repeated string address = 2;
    string toaddr = 3 ;
    uint64 time               = 4;
    uint64  type = 5;
}

message SDKUtxo
{
	string address = 1;
    string hash = 2; // utxo hash
    uint64 value = 3; // utxo value
    uint32 n = 4; // utxo n
}

message SDKBonusamout
{
    string address = 1;
    uint64 invest_amount = 2;
}

message SDKBonusTx
{
    string address = 1;
    string utxo = 2;
    bytes tx = 3;  
}

message SDKPledgeTx
{
    string address = 1;
    bytes utxo = 2;
    bytes tx =3;
}

message Claimtx
{
    string address = 1;
    string utxo = 2;
    bytes tx =3;
}


message AbnormalAddrCnt
{
    string address = 1;
    uint64 count = 2;
}

message SDKClaimBonusTx
{
    string address = 1;
    string utxo = 2;
    bytes tx = 3;  
}

message SDKNodeInfo 
{
  bytes pub = 1;
  bytes sign = 2;
  bytes identity = 3;
  string base58addr = 4;
  string public_base58addr = 5;
  string name	= 6;
   string logo	= 7;
  uint32 listen_ip = 8;
  uint32 listen_port = 9;
  uint32 public_ip = 10;
  uint32 public_port = 11;  
  uint32 height = 12;
  uint64 time_stamp = 13;
  string version = 14;
}


message GetSDKAck
{
    string version = 1;
    int32 code = 2;
    string message = 3;

    uint64 height = 4;
    repeated SDKUtxo utxos = 5;
    repeated bytes  blocks  = 6;
    repeated SDKBonusamout bonusamount = 8;  
    repeated SDKPledgeTx pledgetx = 9;  
    repeated string pledgeaddr = 10; 
    repeated string pledgeutxo = 11; 
    repeated string BonusAddr = 12; 
    repeated string InvestedAddr = 13; 
    repeated SDKBonusTx  BonusTx = 14;  
    repeated string  BonusAddrInvestUtxos = 15;

  
    repeated Claimtx claimtx = 16;
    repeated  AbnormalAddrCnt  abnormaladdr_cnt= 17;

    uint64   M2 = 18;
    uint64  Totalinvest = 19;
    uint64  TotalAwardAmount = 20;

    repeated  SDKNodeInfo   nodeinfo = 22;
    uint64  type = 23;

    repeated SDKClaimBonusTx  ClaimBonusTx = 24;
    repeated string ClaimInvestedAddr = 26;
    repeated string  ClaimBonusAddrInvestUtxos = 27;
}

enum vrfAgentType
{
    vrfAgentType_defalut = 0; 
    vrfAgentType_vrf  = 1;
    vrfAgentType_local = 2;
    vrfAgentType_unknow =3;
};

message SDKContractDeployReq
{
    string   version = 1;  
    string   address = 2;      
    string   deploy_amount = 3; 
    int32    nContractType  = 4;  
    string   pubStr = 5;
}

message SDKContractDeployAck
{
    string version = 1; 
    int32 code = 2; 
    string message = 3; 
    bytes tx = 4;
    uint64  top = 5;
    vrfAgentType isNeedAgent_flag = 8;
    Vrf      vrfInfo              = 9;
}


message SDKContractCallReq
{
    string   version = 1;        
    string   fromaddr = 2;  
    string   toaddr = 3;   
    string   deploy_strTxHash = 4;
}

message SDKContractCallAck
{
    string version = 1; 
    int32 code = 2;
    string message = 3; 
    bytes tx = 4;
    uint64  top = 5;
    vrfAgentType isNeedAgent_flag = 8;
    Vrf      vrfInfo              = 9;
}